File: test/helpers-test.ts
import { trimUndefined, trimUndefinedRecursively } from '../src/helpers'

function assert(condition: boolean) {
  if (!condition) {
    console.error(new Error('Assertion failed'))
    process.exit(1)
  }
}

const user: any = {
  name: 'Alice',
  role: undefined,
  after_undefined: undefined,
  last: 'value',
}
trimUndefined(user)
if ('role' in user) {
  console.error('undefined field is not removed')
  process.exit(1)
}
if ('after_undefined' in user) {
  console.error('continuous undefined fields are not removed')
  process.exit(1)
}
assert(user.name === 'Alice')
assert(user.last === 'value')

const a = {
  name: 'a',
  b: undefined as any,
  extra: undefined,
}
const b = {
  name: 'b',
  a,
}
a.b = b

trimUndefinedRecursively(b)
if ('extra' in b.a) {
  console.error('undefined field is not removed recursively')
  process.exit(1)
}

assert(a.name === 'a')
assert(a.b === b)
assert(b.name === 'b')
assert(b.a === a)

console.log('pass: helpers.ts')

File: python/src/compress_json/helpers_test.py
from helpers import trim_undefined, trim_undefined_recursively


def assertion(condition: bool):
  if not condition:
    raise Exception("Assertion failed")

user = {
  'name': 'Alice',
  'role': None,
  'after_undefined': None,
  'last': 'value',
}
trim_undefined(user)
if 'role' in user:
  print("undefined field is not removed")
  exit(1)
if 'after_undefined' in user:
  print("continuous undefined fields are not removed")
  exit(1)
assertion(user['name'] == 'Alice')
assertion(user['last'] == 'value')

a = {
  'name': 'a',
  'b': None,
  'extra': None 
}
b = {
  'name': 'b',
  'a': a
}
a['b'] = b

trim_undefined_recursively(b)
if 'extra' in b['a']:
  print("undefined field is not removed recursively")
  exit(1)

assertion(a['name'] == 'a')
assertion(a['b'] == b)
assertion(b['name'] == 'b')
assertion(b['a'] == a)

print("pass: helpers.py")
